pkg/containerd/events.go

```diff
@@ -73,59 +84,74 @@ func (e *EventMonitor) MonitorPorts(ctx context.Context) {
+					log.Errorf("failed to create port mapping from container create event: %v", err)
 				}
 
-				if err = e.portTracker.Add(ccEvent.ID, ports); err != nil {
-					log.Errorf("adding port mapping to tracker failed: %w", err)
+				err = e.portTracker.Add(ccEvent.ID, ports)
+				if err != nil {
+					log.Errorf("adding port mapping to tracker failed: %v", err)
+
+					continue
 				}
+
+				updateListener(ctx, ports, e.listenerTracker.Add)
+
 			case "/containers/update":
 				cuEvent := &events.ContainerUpdate{}
 				err := proto.Unmarshal(envelope.Event.Value, cuEvent)
 				if err != nil {
-					log.Errorf("failed unmarshaling container update event: %w", err)
+					log.Errorf("failed unmarshaling container update event: %v", err)
 				}
 
 				ports, err := e.createPortMapping(ctx, cuEvent.ID)
 				if err != nil {
-					log.Errorf("failed to create port mapping from container update event: %w", err)
+					log.Errorf("failed to create port mapping from container update event: %v", err)
 				}
 
 				existingPortMap := e.portTracker.Get(cuEvent.ID)
 				if existingPortMap != nil {
 					if !reflect.DeepEqual(ports, existingPortMap) {
 						err := e.portTracker.Remove(cuEvent.ID)
 						if err != nil {
-							log.Errorf("failed to remove port mapping from container update event: %w", err)
-							err := e.portTracker.Add(cuEvent.ID, ports)
-							if err != nil {
-								log.Errorf("failed to add port mapping from container update event: %w", err)
-							}
+							log.Errorf("failed to remove port mapping from container update event: %v", err)
+						}
+
+						updateListener(ctx, ports, e.listenerTracker.Remove)
+						err = e.portTracker.Add(cuEvent.ID, ports)
+						if err != nil {
+							log.Errorf("failed to add port mapping from container update event: %v", err)
 
 							continue
 						}
+
+						updateListener(ctx, ports, e.listenerTracker.Add)
 					}
 				}
 				// Not 100% sure if we ever get here...
 				if err = e.portTracker.Add(cuEvent.ID, ports); err != nil {
-					log.Errorf("failed to add port mapping from container update event: %w", err)
+					log.Errorf("failed to add port mapping from container update event: %v", err)
 				}
 
 			case "/containers/delete":
 				cdEvent := &events.ContainerDelete{}
 				err := proto.Unmarshal(envelope.Event.Value, cdEvent)
 				if err != nil {
-					log.Errorf("failed unmarshaling container delete event: %w", err)
+					log.Errorf("failed unmarshaling container delete event: %v", err)
 				}
 
-				err = e.portTracker.Remove(cdEvent.ID)
-				if err != nil {
-					log.Errorf("removing port mapping from tracker failed: %w", err)
+				portMapToDelete := e.portTracker.Get(cdEvent.ID)
+				if portMapToDelete != nil {
+					err = e.portTracker.Remove(cdEvent.ID)
+					if err != nil {
+						log.Errorf("removing port mapping from tracker failed: %v", err)
+					}
+
+					updateListener(ctx, portMapToDelete, e.listenerTracker.Remove)
```

Comment:
Why are we not calling this if `portMapToDelete` is `nil`? This doesn't match the update case (where we always remove the port even if there's no port map).  I think that was better.